File: src/services/budgetCentreProjectCodeService.ts
import { authService } from './authService';

const API_BASE_URL = "http://localhost:7080/api";

export interface BudgetCentreProjectCodeRequest {
  centreProjectCode: string;
  centreProject: string;
  budgetCentreProjectFullName: string;
  budgetCentreProjectShortName: string;
  fromDate: string;
  toDate?: string | null;
  userId: string;
  regStatus: string;
}

export interface BudgetCentreProjectCodeResponse extends BudgetCentreProjectCodeRequest {
  active: boolean;
}

class BudgetCentreProjectCodeService {
  private getHeaders(): Record<string, string> {
    const authHeaders = authService.getAuthHeader();
    
    return {
      "Content-Type": "application/json",
      ...authHeaders
    };
  }

  async getAllBudgetCentreProjectCodes(): Promise<BudgetCentreProjectCodeResponse[]> {
    try {
      const response = await fetch(`${API_BASE_URL}/budget-centre-project-codes`, {
        method: "GET",
        headers: this.getHeaders(),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to fetch budget centre project codes`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching budget centre project codes:", error);
      throw error;
    }
  }

  async getActiveBudgetCentreProjectCodes(): Promise<BudgetCentreProjectCodeResponse[]> {
    try {
      const response = await fetch(`${API_BASE_URL}/budget-centre-project-codes/active`, {
        method: "GET",
        headers: this.getHeaders(),
        
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to fetch active budget centre project codes`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching active budget centre project codes:", error);
      throw error;
    }
  }

  async getInactiveBudgetCentreProjectCodes(): Promise<BudgetCentreProjectCodeResponse[]> {
    try {
      const response = await fetch(`${API_BASE_URL}/budget-centre-project-codes/inactive`, {
        method: "GET",
        headers: this.getHeaders(),
        
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to fetch inactive budget centre project codes`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching inactive budget centre project codes:", error);
      throw error;
    }
  }

  async getBudgetCentreProjectCodeByKey(centreProjectCode: string, centreProject: string): Promise<BudgetCentreProjectCodeResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/budget-centre-project-codes/${centreProjectCode}/${centreProject}`, {
        method: "GET",
        headers: this.getHeaders(),
        
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to fetch budget centre project code`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching budget centre project code:", error);
      throw error;
    }
  }

  async createBudgetCentreProjectCode(request: BudgetCentreProjectCodeRequest): Promise<BudgetCentreProjectCodeResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/budget-centre-project-codes`, {
        method: "POST",
        headers: this.getHeaders(),
        
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to create budget centre project code`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error creating budget centre project code:", error);
      throw error;
    }
  }

  async updateBudgetCentreProjectCode(centreProjectCode: string, centreProject: string, request: BudgetCentreProjectCodeRequest): Promise<BudgetCentreProjectCodeResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/budget-centre-project-codes/${centreProjectCode}/${centreProject}`, {
        method: "PUT",
        headers: this.getHeaders(),
        
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to update budget centre project code`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error updating budget centre project code:", error);
      throw error;
    }
  }

  async deactivateBudgetCentreProjectCode(centreProjectCode: string, centreProject: string): Promise<BudgetCentreProjectCodeResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/budget-centre-project-codes/${centreProjectCode}/${centreProject}/deactivate`, {
        method: "PUT",
        headers: this.getHeaders(),
        
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to deactivate budget centre project code`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error deactivating budget centre project code:", error);
      throw error;
    }
  }

  async deleteBudgetCentreProjectCode(centreProjectCode: string, centreProject: string): Promise<void> {
    try {
      const response = await fetch(`${API_BASE_URL}/budget-centre-project-codes/${centreProjectCode}/${centreProject}`, {
        method: "DELETE",
        headers: this.getHeaders(),
        
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to delete budget centre project code`);
      }
    } catch (error) {
      console.error("Error deleting budget centre project code:", error);
      throw error;
    }
  }
}

export const budgetCentreProjectCodeService = new BudgetCentreProjectCodeService();

File: pms-backend/src/main/java/com/pms/service/BudgetCentreProjectCodeService.java
package com.pms.service;

import com.pms.dto.BudgetCentreProjectCodeRequest;
import com.pms.dto.BudgetCentreProjectCodeResponse;
import com.pms.entity.BudgetCentreProjectCode;
import com.pms.entity.BudgetCentreProjectCodeId;
import com.pms.repository.BudgetCentreProjectCodeRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class BudgetCentreProjectCodeService {
    
    private final BudgetCentreProjectCodeRepository budgetCentreProjectCodeRepository;
    
    // Get all budget centre project codes
    public List<BudgetCentreProjectCodeResponse> getAllBudgetCentreProjectCodes() {
        return budgetCentreProjectCodeRepository.findAllByOrderByCentreProjectCodeAsc()
                .stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }
    
    // Get only active codes
    public List<BudgetCentreProjectCodeResponse> getActiveBudgetCentreProjectCodes() {
        return budgetCentreProjectCodeRepository.findAllActive()
                .stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }
    
    // Get only inactive codes
    public List<BudgetCentreProjectCodeResponse> getInactiveBudgetCentreProjectCodes() {
        return budgetCentreProjectCodeRepository.findAllInactive()
                .stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }
    
    // Get by composite key (centreProjectCode + centreProject)
    public BudgetCentreProjectCodeResponse getBudgetCentreProjectCodeByKey(String centreProjectCode, String centreProject) {
        BudgetCentreProjectCode code = budgetCentreProjectCodeRepository
                .findByCentreProjectCodeAndCentreProject(centreProjectCode, centreProject)
                .orElseThrow(() -> new RuntimeException("Budget Centre Project Code not found with code: " + centreProjectCode + " and project: " + centreProject));
        return convertToResponse(code);
    }
    
    // Create new budget centre project code
    @Transactional
    public BudgetCentreProjectCodeResponse createBudgetCentreProjectCode(BudgetCentreProjectCodeRequest request) {
        // Validation
        validateBudgetCentreProjectCodeRequest(request);
        
        // Check if code already exists
        BudgetCentreProjectCodeId id = new BudgetCentreProjectCodeId(request.getCentreProjectCode(), request.getCentreProject());
        if (budgetCentreProjectCodeRepository.existsById(id)) {
            throw new RuntimeException("Budget Centre Project Code with code " + request.getCentreProjectCode() + " and project " + request.getCentreProject() + " already exists");
        }
        
        // Check if short name is unique
        if (budgetCentreProjectCodeRepository.findByBudgetCentreProjectShortName(request.getBudgetCentreProjectShortName()).isPresent()) {
            throw new RuntimeException("Budget Centre Project Code with short name " + request.getBudgetCentreProjectShortName() + " already exists");
        }
        
        BudgetCentreProjectCode code = BudgetCentreProjectCode.builder()
                .centreProjectCode(request.getCentreProjectCode())
                .centreProject(request.getCentreProject())
                .budgetCentreProjectFullName(request.getBudgetCentreProjectFullName())
                .budgetCentreProjectShortName(request.getBudgetCentreProjectShortName())
                .fromDate(request.getFromDate())
                .toDate(request.getToDate())
                .userId(request.getUserId())
                .regStatus(request.getRegStatus())
                .regTime(LocalDate.now())
                .build();
        
        BudgetCentreProjectCode savedCode = budgetCentreProjectCodeRepository.save(code);
        log.info("Budget Centre Project Code created successfully: {}/{}", savedCode.getCentreProjectCode(), savedCode.getCentreProject());
        
        return convertToResponse(savedCode);
    }
    
    // Update budget centre project code
    @Transactional
    public BudgetCentreProjectCodeResponse updateBudgetCentreProjectCode(String centreProjectCode, String centreProject, BudgetCentreProjectCodeRequest request) {
        BudgetCentreProjectCode code = budgetCentreProjectCodeRepository
                .findByCentreProjectCodeAndCentreProject(centreProjectCode, centreProject)
                .orElseThrow(() -> new RuntimeException("Budget Centre Project Code not found with code: " + centreProjectCode + " and project: " + centreProject));
        
        // Validation
        validateBudgetCentreProjectCodeRequest(request);
        
        // Check if short name is already used by another code
        budgetCentreProjectCodeRepository.findByBudgetCentreProjectShortName(request.getBudgetCentreProjectShortName())
                .ifPresent(existing -> {
                    if (!existing.getCentreProjectCode().equals(centreProjectCode) || !existing.getCentreProject().equals(centreProject)) {
                        throw new RuntimeException("Budget Centre Project Code with short name " + request.getBudgetCentreProjectShortName() + " already exists");
                    }
                });
        
        code.setBudgetCentreProjectFullName(request.getBudgetCentreProjectFullName());
        code.setBudgetCentreProjectShortName(request.getBudgetCentreProjectShortName());
        code.setFromDate(request.getFromDate());
        code.setToDate(request.getToDate());
        code.setUserId(request.getUserId());
        code.setRegStatus(request.getRegStatus());
        
        BudgetCentreProjectCode updatedCode = budgetCentreProjectCodeRepository.save(code);
        log.info("Budget Centre Project Code updated successfully: {}/{}", updatedCode.getCentreProjectCode(), updatedCode.getCentreProject());
        
        return convertToResponse(updatedCode);
    }
    
    // Deactivate budget centre project code (set toDate)
    @Transactional
    public BudgetCentreProjectCodeResponse deactivateBudgetCentreProjectCode(String centreProjectCode, String centreProject) {
        BudgetCentreProjectCode code = budgetCentreProjectCodeRepository
                .findByCentreProjectCodeAndCentreProject(centreProjectCode, centreProject)
                .orElseThrow(() -> new RuntimeException("Budget Centre Project Code not found with code: " + centreProjectCode + " and project: " + centreProject));
        
        code.setToDate(LocalDate.now());
        BudgetCentreProjectCode deactivated = budgetCentreProjectCodeRepository.save(code);
        
        log.info("Budget Centre Project Code deactivated: {}/{}", centreProjectCode, centreProject);
        return convertToResponse(deactivated);
    }
    
    // Delete budget centre project code
    @Transactional
    public void deleteBudgetCentreProjectCode(String centreProjectCode, String centreProject) {
        BudgetCentreProjectCode code = budgetCentreProjectCodeRepository
                .findByCentreProjectCodeAndCentreProject(centreProjectCode, centreProject)
                .orElseThrow(() -> new RuntimeException("Budget Centre Project Code not found with code: " + centreProjectCode + " and project: " + centreProject));
        
        budgetCentreProjectCodeRepository.delete(code);
        log.info("Budget Centre Project Code deleted: {}/{}", centreProjectCode, centreProject);
    }
    
    // Validation method
    private void validateBudgetCentreProjectCodeRequest(BudgetCentreProjectCodeRequest request) {
        if (request.getCentreProjectCode() == null || request.getCentreProjectCode().trim().isEmpty()) {
            throw new RuntimeException("Centre Project Code is required");
        }
        
        if (request.getCentreProjectCode().length() > 2) {
            throw new RuntimeException("Centre Project Code cannot exceed 2 characters");
        }
        
        if (request.getCentreProject() == null || request.getCentreProject().trim().isEmpty()) {
            throw new RuntimeException("Centre Project is required");
        }
        
        if (request.getCentreProject().length() > 1) {
            throw new RuntimeException("Centre Project must be a single character");
        }
        
        if (request.getBudgetCentreProjectFullName() == null || request.getBudgetCentreProjectFullName().trim().isEmpty()) {
            throw new RuntimeException("Budget Centre Project Full Name is required");
        }
        
        if (request.getBudgetCentreProjectShortName() == null || request.getBudgetCentreProjectShortName().trim().isEmpty()) {
            throw new RuntimeException("Budget Centre Project Short Name is required");
        }
        
        if (request.getFromDate() == null) {
            throw new RuntimeException("From Date is required");
        }
        
        if (request.getToDate() != null && request.getToDate().isBefore(request.getFromDate())) {
            throw new RuntimeException("To Date cannot be before From Date");
        }
        
        if (request.getUserId() == null || request.getUserId().trim().isEmpty()) {
            throw new RuntimeException("User ID is required");
        }
        
        if (request.getRegStatus() == null || request.getRegStatus().trim().isEmpty()) {
            throw new RuntimeException("Registration Status is required");
        }
        
        if (!request.getRegStatus().matches("[A-Za-z]")) {
            throw new RuntimeException("Registration Status must be a single character");
        }
    }
    
    private BudgetCentreProjectCodeResponse convertToResponse(BudgetCentreProjectCode code) {
        return BudgetCentreProjectCodeResponse.builder()
                .centreProjectCode(code.getCentreProjectCode())
                .centreProject(code.getCentreProject())
                .budgetCentreProjectFullName(code.getBudgetCentreProjectFullName())
                .budgetCentreProjectShortName(code.getBudgetCentreProjectShortName())
                .fromDate(code.getFromDate())
                .toDate(code.getToDate())
                .userId(code.getUserId())
                .regStatus(code.getRegStatus())
                .active(code.isActive())
                .build();
    }
}
